Create MailListener v2

hod-alpert opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/26326
fixes: https://github.com/demisto/etc/issues/19920
fixes: https://github.com/demisto/etc/issues/20089
fixes: https://github.com/demisto/etc/issues/14581

## Description
Added a v2 version of Mail listener
Regarding [this bug](https://github.com/demisto/etc/issues/26326) - in case the mail does not have the `RFC822` header it will not be fetched at all.
Regarding [this bug](https://github.com/demisto/etc/issues/19920)- both Japanise and Russian were tested and no problem occured during the parsing of mail containing content from those languages (see screenshot attached)

Using IMAPClient package
## Screenshots
![image](https://user-images.githubusercontent.com/41257953/89124698-a0479f80-d4e1-11ea-86a2-e8b21675add6.png)

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [ ] 5.5.0
- [x] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [ ] Documentation

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,81 @@
+  name: port
+  required: true
+  type: 0
+- name: credentials
```
set `display` here to `Username`

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,81 @@
+  name: incidentType
+  required: false
+  type: 13
+- display: IMAP port
```
default should be `143` no ?

Itay4 replied:
```suggestion
- display: IMAP Port
```

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,81 @@
+  id: Mail Listener v2
+  version: -1
+configuration:
+- display: Mail server hostname or IP address
```
```suggestion
- display: Mail Server Hostname / IP address
```

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,81 @@
+  name: insecure
+  required: false
+  type: 8
+- display: Use system proxy settings
```
add todo here to remove in case we decide to

Itay4 commented on Packs/MailListener/pack_metadata.json:
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "MailListenerV2",
```
```suggestion
    "name": "Mail Listener",
```

Itay4 replied:
what about that?

Itay4 commented on Packs/MailListener/pack_metadata.json:
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "MailListenerV2",
+    "description": "MailListenerV2",
```
improve description here

Itay4 commented on Tests/test_integration.py:
```diff
@@ -641,7 +641,7 @@ def __create_incident_with_playbook(client, name, playbook_id, integrations, pri
 
     try:
-        response = client.create_incident(create_incident_request=create_incident_request)
+        response = client.convert_to_incident(create_incident_request=create_incident_request)
```
why the change here?

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+import dateparser
+from datetime import timezone
+
+# CONSTS
```
lets remove that comment

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+from typing import Dict, Any
```
please run `isort` here

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+    messages = client.search(messages_query)
+    messages = messages[:50]
+    mails_fetched = []
+    for uid, message_data in client.fetch(messages, 'RFC822').items():
```
`uid` is not used here, perhaps you can use `values()` instead of `items()` ?

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+
+from imapclient import IMAPClient
+from imap_tools import OR
+import mailparser
```
if we need only `parse_from_bytes` so lets import only that

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+import mailparser
+import demistomock as demisto
+from CommonServerPython import *  # noqa: E402 lgtm [py/polluting-import]
+import dateparser
```
if we need only `parse` lets import just that

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+from datetime import timezone
+
+# CONSTS
+INCLUDE_RAW_BODY = demisto.params().get('Include_raw_body', False)
```
please move all params to initialized in the `main` func

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+
+    LOG(f'Command being called is {demisto.command()}')
+    try:
+        with IMAPClient(mail_server_url, ssl=verify_ssl, timeout=20, port=port) as client:
```
why is `timeout` set to `20` ?

hod-alpert replied:
Some reasonable time, if this is not specified there is no timeout at all

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+
+
+def test_module(client: IMAPClient) -> str:
+    client.search()
```
can we narrow the search ?
im a bit concerned this might take long

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+        return raw_json
+
+
+def fetch_incidents(client,
```
please add type hints

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+    Args:
+        client (Client): HelloWorld client
+        last_run (dateparser.time): The greatest incident created_time we fetched from last fetch
+        first_fetch_time (dateparser.time): If last_run is None then fetch all incidents since first_fetch_time
```
```suggestion
        first_fetch_time (str): If last_run is None then fetch all incidents since first_fetch_time
```

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+
+    Args:
+        client (Client): HelloWorld client
+        last_run (dateparser.time): The greatest incident created_time we fetched from last fetch
```
```suggestion
        last_run (dict): The greatest incident created_time we fetched from last fetch
```

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+    This function will execute each interval (default is 1 minute).
+
+    Args:
+        client (Client): HelloWorld client
```
update docstring here

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+    return next_run, incidents
+
+
+def generate_search_query(latest_created_time, permitted_from_addresses, permitted_from_domains):
```
please add type hints and docstring

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+        latest_created_time = datetime.fromisoformat(last_fetch)
+    messages_query = generate_search_query(latest_created_time, PERMITTED_FROM_ADDRESSES, PERMITTED_FROM_DOMAINS)
+    messages = client.search(messages_query)
+    messages = messages[:50]
```
the limit should not hard coded value, but an integration param

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+            client.login(username, password)
+            client.select_folder(folder)
+            if demisto.command() == 'test-module':
+                # This is the call made when pressing the integration Test button.
```
remove this comment please

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+                demisto.results(result)
+
+            elif demisto.command() == 'fetch-incidents':
+                # Set and define the fetch incidents command to run after activated via integration settings.
```
remove this comment please

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+                demisto.setLastRun(next_run)
+                demisto.incidents(incidents)
+    # Log exceptions
+    except Exception as e:
```
remove this comment please

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+                demisto.incidents(incidents)
+    # Log exceptions
+    except Exception as e:
+        LOG(e)
```
can remove `LOG` from here as `return_error` already calls it

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+    folder = params.get('folder')
+    username = demisto.params().get('credentials').get('identifier')
+    password = demisto.params().get('credentials').get('password')
+    verify_ssl = not params.get("insecure", False)
```
```suggestion
    verify_ssl = not params.get('insecure', False)
```

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+
+
+def main():
+    """
```
remove this comment please

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+    messages = client.search(messages_query)
+    messages = messages[:50]
+    mails_fetched = []
+    for uid, message_data in client.fetch(messages, 'RFC822').items():
```
why `RFC822` ?

hod-alpert replied:
From what I see it seems to be the gold standard format to send mails in.

Do you know of any issue that might occur as a result of this?

Itay4 replied:
what other options do we have here? why not just get them all?

hod-alpert replied:
The way I see it this is the format in which we want to fetch, but the emails returned are the same emails (At least that's the results from checking on my own email)
The other options I saw so far are `INTERNALDATE` which returns only the date in which the mail was recived and `ENVELOPE` which fetches the `'bcc','cc','count','date','from_','in_reply_to','index','message_id','reply_to','sender','subject','to'` fields bot not the body nor the attachments.

I seem like this is the format that brings all the data we need in the mail.

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+    messages = messages[:50]
+    mails_fetched = []
+    for uid, message_data in client.fetch(messages, 'RFC822').items():
+        message_bytes = message_data.get(b'RFC822')
```
why are we working with bytes here? what other options we got?

hod-alpert replied:
None the keys we have with message_data are `b'SEQ` which is a number and the bytes data.
Why is it matter?

Itay4 replied:
just wondering.. if that's the only option then ok

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+            mails_fetched.append(email_message_object)
+    if mails_fetched:
+        latest_created_time = max(mails_fetched, key=lambda x: x.date).date
+    incidents = [mail.convert_to_incident() for mail in mails_fetched]
```
feels to me like another redundant loop, we can do it in the loop in which we build `mails_fetched`

hod-alpert replied:
Same

Itay4 replied:
ok your call

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+        if email_message_object.date > latest_created_time:
+            mails_fetched.append(email_message_object)
+    if mails_fetched:
+        latest_created_time = max(mails_fetched, key=lambda x: x.date).date
```
feels to me like another redundant loop, we can do it in the loop in which we build `mails_fetched`

hod-alpert replied:
We could but that will uglify the code because it will have to happen under the `if email_message_object.date > latest_created_time`.

As long as we have a limit the loop should be pretty fast and we get a much readable code IMHO.

Itay4 replied:
ok your call

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,193 @@
+            # check for error
+            if file_result['Type'] == entryTypes['error']:
+                demisto.error(file_result['Contents'])
+                raise Exception(file_result['Contents'])
```
not sure we want to raise an exception here

hod-alpert replied:
You think demisto.error is enough?

Itay4 replied:
yes

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,220 @@
+from imapclient import IMAPClient
+
+import demistomock as demisto
+from CommonServerPython import *  # noqa: E402 lgtm [py/polluting-import]
```
i dont think we need `# noqa: E402 lgtm [py/polluting-import]`
please check if lint passes if you remove it

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,90 @@
+  name: first_fetch
+  required: false
+  type: 0
+- display: Trust any certificate (not secure)
```
please move the insecure and proxy params to be below all other params

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,220 @@
+        return raw_json
+
+
+def fetch_incidents(client: IMAPClient,
```
please add return type hint

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,220 @@
+    This function will execute each interval (default is 1 minute).
+
+    Args:
+        client: HelloWorld client
```
please update `client` description

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,220 @@
+    return next_run, incidents
+
+
+def generate_search_query(latest_created_time: datetime, permitted_from_addresses: str, permitted_from_domains: str):
```
please return type hint

Itay4 replied:
please unit test for that one

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,220 @@
+
+
+def generate_search_query(latest_created_time: datetime, permitted_from_addresses: str, permitted_from_domains: str):
+    """
```
please add to the docstring what the function does, an example, and fill in the return section

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,92 @@
+  name: permittedFromAdd
+  required: false
+  type: 0
+- display: Forward mails from these domains only (eg. demo.com, test.com)
```
improve as discussed

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,92 @@
+  name: limit
+  required: false
+  type: 0
+- display: Trust any certificate (not secure)
```
add tls param

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,92 @@
+  required: false
+  type: 8
+#TODO think if we want to remove
+- display: Use system proxy settings
```
remove proxy

Itay4 commented:
after merging this, lets deprecate v1

hod-alpert commented:
@Itay4 Fixed all comments and tested on exchange server with non-English folder.

Added the labels again since some were missing for the playbook.

Added the list command

Added the TLS parameter and modified the implementation of the insecure param (works with unverified certificate)

The only thing that's left to do is pass the phishing playbook which fails on something I don't understand, I'll consult with @idovandijk tomorrow about it.

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,348 @@
+    username = demisto.params().get('credentials').get('identifier')
+    password = demisto.params().get('credentials').get('password')
+    verify_ssl = not params.get('insecure', False)
+    TLS_connection = params.get('TLS_connection', True)
```
```suggestion
    tls_connection = params.get('TLS_connection', True)
```

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,113 @@
+  name: save_file
+  required: false
+  type: 8
+- display: Use TLS for connection (defaults to True)
```
why `defaults to True` ?

hod-alpert replied:
No particular reason, seems like a better default then False, remove it?

Itay4 replied:
yes, like `insecure`

Itay4 commented:
please generate integration readme

Itay4 commented:
> This pull request **introduces 1 alert** when merging [850cb35](https://github.com/demisto/content/commit/850cb35e7c4653b71315a276e04151f9a6d896c9) into [ed7279b](https://github.com/demisto/content/commit/ed7279bd34e2a1e854073f308f2f9077ded629cc) - [view on LGTM.com](https://lgtm.com/projects/g/demisto/content/rev/pr-ee70a286bb9c2c6e0ea35ddf21c02e106b60dfdf)
> 
> **new alerts:**
> 
> * 1 for Unused import

@hod-alpert please handle the alert

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,348 @@
+                           'value': ','.join([attachment['filename'] for attachment in self.attachments])})
+        return labels
+
+    def parse_attachments(self):
```
please add return type hint

Itay4 commented on Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:
```diff
@@ -32,8 +32,8 @@
-CONTENT_RELEASE_VERSION = '0.0.0'
-CONTENT_BRANCH_NAME = 'master'
+CONTENT_RELEASE_VERSION = '2.5.0'
+CONTENT_BRANCH_NAME = 'hod/mail_listener'
```
revert that

hod-alpert replied:
Oops, didn't see that

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,162 @@
+  name: MailServerURL
+  required: true
+  type: 0
+- display: Fetch incidents
```
please place the all fetch related params together, and make sure they are below the credentials

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/README.md:
```diff
@@ -0,0 +1,203 @@
+
+---
+
+## Configure MAil Listener v2 on XSOAR
```
```suggestion
## Configure Mail Listener v2 on XSOAR
```

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/README.md:
````diff
@@ -0,0 +1,203 @@
+}
+```
+
+#### Human Readable Output
````
remove this section

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,162 @@
+  - deprecated: false
+    description: Fetches mails according to the configuration
+    execution: false
+    name: mail-listener-list-messages
```
i dont like that this command uses the terminology `messages` while the other use `email`
lets set them all to `email` or `message` and lets make sure its according to the standard `mail-listener-email/message-...`

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.yml:
```diff
@@ -0,0 +1,162 @@
+    execution: false
+    name: mail-listener-list-messages
+    outputs:
+    - contextPath: MailListener.EmailPreview.Subject
```
need to make sure we output to the context path, `Email` and not `EmailPreview`

hod-alpert replied:
But the objects returned are not the same.
In list only `Subject`, `To`, `From`, `Date` and `ID` are returned.
In get email command the full email is returned.
Since they are not the same I thought is best to put them in different path in the context.

Itay4 replied:
if the same data is populated then it should be the same path, the number of fields returned does not really matter

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2_description.md:
```diff
@@ -0,0 +1 @@
+Listens to a mailbox and enables incident triggering via e-mail.
```
here should come only detailed instructions, not the integration description

hod-alpert replied:
There is no detailed instructions, what should I put there?
The configuration's instructions? like below?

## Configure Mail Listener v2 on XSOAR

---

1. Navigate to __Settings__ > __Integrations__ > __Servers & Services__.
2. Search for Mail Listener v2. 
3. Click __Add instance__ to create and configure a new integration instance.
    * __Name__: a textual name for the integration instance.
    * __Fetch incidents__: Whether to fetch incidents or not
    * __MailServerURL__: Mail Server Hostname / IP address
    * __port__: IMAP Port
    * __credentials__: Username and password
    * __folder__: Incoming mail folder
    * __permittedFromAdd__: Fetch mails from these senders addresses only (eg. admin@demo.com,test@demo.com)
    * __first_fetch__: First fetch time (\<number\> \<time unit\>, e.g., 12 hours, 7 days, 3 months, 1 year)
    * __limit__: The maximum number of incidents to fetch each time
    * __delete_processed__: Delete processed emails
    * __Include_raw_body__: Include raw body in incidents
    * __save_file__: Save the email .eml file
    * __TLS_connection__: Use TLS for connection (defaults to True)
    * __insecure__: Trust any certificate (not secure)
    * __incidentFetchInterval__: Incidents Fetch Interval
4. Click __Test__ to validate the connection and the authentication.

Itay4 replied:
if there isnt, just keep it empty

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,393 @@
+    return 'ok'
+
+
+def list_messages(client, first_fetch_time, permitted_from_addresses, permitted_from_domains):
```
please make sure we have type hints for all functions

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,393 @@
+                save_file: bool = False,
+                message_id: int = None):
+    """
+    This function will fetch the mails from the imap servers.
```
```suggestion
    This function will fetch the mails from the IMAP server.
```

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,393 @@
+    if message_id:
+        messages = [message_id]
+    else:
+        messages_query = generate_search_query(first_fetch_time,  # type: ignore[arg-type]
```
we can avoid ignoring mypy validation here by handling better `first_fetch_time` type

Itay4 commented on Packs/MailListener/Integrations/MailListenerV2/MailListenerV2.py:
```diff
@@ -0,0 +1,408 @@
+        # Removing Parenthesis and quotes
+        messages_query = messages_query.strip('()').replace('"', '')
+    # Creating a list of the OR query words
+    messages_query_list = messages_query.split() + ['SINCE', latest_created_time]  # type: ignore[list-item]
```
can we convert `latest_created_time` to str and remove the type ignore ?

hod-alpert replied:
No, it must be a datetime object

Itay4 commented on Packs/MailListener/pack_metadata.json:
```diff
@@ -0,0 +1,27 @@
+    "tags": [],
+    "useCases": [],
+    "keywords": [],
+        "dependencies": {
```
fix identation here
why did you set dependencies ?

hod-alpert commented:
@kirbles19 Added additional commands and a README file.
Could you re-review?

richardbluestone commented:
I've reviewed and updated the yml file. The markdown file is the generated version of the yml file, so no need to review, unless there is something else you want me to review?

hod-alpert commented:
@richardbluestone No, thanks a lot
